src/posts/2020-07-30-the-worst-decision-of-my-career.md

```diff
@@ -0,0 +1,271 @@
+my lessons, I've studied the books, and I even met some of my gurus that wrote
+them! This time I'll follow "industry standards"!_
+
+_I am the "architect"; I will design the perfect system! Without me, none of
+this will be possible. Those mindless programmers have no idea what they are
+doing. I, and only I, am the true heir of Martin Fowler!_
+
+This was a bit dramatic, but I needed a break from all of that whining. I know
+it's easy to fall into these traps. It's even easier when your company raised
+money, and everyone is expecting you to deliver the absolute best product ever
+because they are paying you for it! This is why your starting team is so
+important; they have to withstand the pressure. They have to know that in order
+to build the grand vision, they have to go one step at a time.
+
+I was going to write a whole section on what makes for a good starting team,
+but that's a different topic, and there's still so much I need to say.
+
+## The decision
+
+I've made a lot of bad decisions, and I've been fortunate enough to suffer the
+consequences of those decisions. A lot of developers don't get to experience
+that, so they never learn. That is also why I'm always skeptical about external
+"expert" consultants; they usually don't stick around to experience the
+suffering.
+
+For many years, we, Subvisual, built web products with some success, so we felt
+confident. As if there was nothing we couldn't do. We've already learned a lot,
+so we know what to avoid. Never again, we said!
+
+So what was the worst decision of my career? It's clear to me - although some
+will disagree - that the worst decision I made was starting a product with
+event-sourcing, asynchronous communications, event-driven architecture, and all
+of that. Mind you, I didn't make the decision alone, but that doesn't change
+anything.
+
+Why was that the worst decision? Because we didn't know the market. We didn't
+know if the market even existed! We didn't know what we were building and for
+whom. We didn't know enough, and the decisions we made crippled our ability to
+change. The business rules were almost set in stone. Migrating data was a pain
+and the source of so many bugs. We didn't start with microservices, but we
+still managed to start the app with more than a dozen umbrella apps (we were
+using Elixir) that, in theory, could be extracted later. And because we didn't
+know any better, the event store, which kept the state of **all** services, was
+also being used as an event-bus to communicate between those services. Because
+of that, it was possible to couple one service to the internal state of
+another, which happened, all over the place, all the time. This almost
+invisible coupling made everything worse.
+
+## Aftermath
+
+As with any new technology (for us), we overdid it at first, and because of
+that we were lost. Every change increased the number of bugs. We spent long
```

Comment:
"over our heads" maybe?

Reply:
I think that also implies that we were going around not really knowing what to do. I both don't think it depicts a true image of what happened and is not something nice for us.

Reply:
But it does, I remember clearly a time when everyone was in Braga, Maciek, Zdenko, Antonin, and we were all talking about this. We were lost. Integration tests were failing randomly and we didn't know why. Commits were breaking the build because of data migrations. The front-end was reporting weird errors. It is a true image. But I understand that we may see things differently. The message I want to convey is that too many things went wrong that we didn't expect. Do you have a more appropriate suggestion?

Reply:
That phrasing is better. Something along the lines of "As with any new tech... there were things that went wrong that we didn't antecipate."

Reply:
👍